File: code_challenge_solution/java/124 Minesweeper.java
import java.util.Random;
import java.util.Scanner;

public class Minesweeper {
    private int row;
    private int col;
    private int numOfMines;
    private char[][] board;
    private boolean[][] revealed;

    public Minesweeper(int row, int col, int numOfMines) {
        this.row = row;
        this.col = col;
        this.numOfMines = numOfMines;
        this.board = new char[row][col];
        this.revealed = new boolean[row][col];
        initializeBoard();
        placeMines();
    }

    private void initializeBoard() {
        for (int i = 0; i < row; i++) {
            for (int j = 0; j < col; j++) {
                board[i][j] = '-';
            }
        }
    }

    private void placeMines() {
        Random random = new Random();
        int count = 0;
        while (count < numOfMines) {
            int x = random.nextInt(row);
            int y = random.nextInt(col);
            if (board[x][y] != '*') {
                board[x][y] = '*';
                count++;
            }
        }
    }

    private void revealCell(int x, int y) {
        if (x < 0 || x >= row || y < 0 || y >= col || revealed[x][y]) {
            return;
        }

        revealed[x][y] = true;

        if (board[x][y] != '-') {
            return;
        }

        int count = countAdjacentMines(x, y);
        if (count > 0) {
            board[x][y] = (char) (count + '0');
        } else {
            board[x][y] = ' ';
            revealCell(x - 1, y - 1);
            revealCell(x - 1, y);
            revealCell(x - 1, y + 1);
            revealCell(x, y - 1);
            revealCell(x, y + 1);
            revealCell(x + 1, y - 1);
            revealCell(x + 1, y);
            revealCell(x + 1, y + 1);
        }
    }

    private int countAdjacentMines(int x, int y) {
        int count = 0;
        for (int i = -1; i <= 1; i++) {
            for (int j = -1; j <= 1; j++) {
                if (i == 0 && j == 0) {
                    continue;
                }
                int newX = x + i;
                int newY = y + j;
                if (newX >= 0 && newX < row && newY >= 0 && newY < col && board[newX][newY] == '*') {
                    count++;
                }
            }
        }
        return count;
    }

    private boolean hasWon() {
        int revealedCount = 0;
        for (int i = 0; i < row; i++) {
            for (int j = 0; j < col; j++) {
                if (revealed[i][j]) {
                    revealedCount++;
                }
            }
        }
        return revealedCount == (row * col) - numOfMines;
    }

    public void play() {
        Scanner scanner = new Scanner(System.in);
        while (true) {
            printBoard();

            System.out.print("Enter row (0-" + (row - 1) + "): ");
            int x = scanner.nextInt();

            System.out.print("Enter column (0-" + (col - 1) + "): ");
            int y = scanner.nextInt();

            if (board[x][y] == '*') {
                System.out.println("Game Over!");
                break;
            }

            revealCell(x, y);

            if (hasWon()) {
                System.out.println("Congratulations! You have won the game!");
                break;
            }
        }
    }

    private void printBoard() {
        System.out.println("Current Board:");
        for (int i = 0; i < row; i++) {
            for (int j = 0; j < col; j++) {
                if (revealed[i][j]) {
                    System.out.print(board[i][j] + " ");
                } else {
                    System.out.print("- ");
                }
            }
            System.out.println();
        }
    }

    public static void main(String[] args) {
        Minesweeper game = new Minesweeper(10, 10, 10);
        game.play();
    }
}

File: code_challenge_solution/python/124 Minesweeper.py
import random

def create_board(rows, cols):
    board = []
    for _ in range(rows):
        row = [' ']*cols
        board.append(row)
    return board

def place_mines(board, num_mines):
    rows = len(board)
    cols = len(board[0])
    mines_placed = 0
    while mines_placed < num_mines:
        row = random.randint(0, rows-1)
        col = random.randint(0, cols-1)
        if board[row][col] != '*':
            board[row][col] = '*'
            mines_placed += 1

def count_adjacent_mines(board, row, col):
    count = 0
    rows = len(board)
    cols = len(board[0])
    for i in range(max(0, row-1), min(rows, row+2)):
        for j in range(max(0, col-1), min(cols, col+2)):
            if board[i][j] == '*':
                count += 1
    return count

def reveal_cell(board, revealed, row, col):
    if revealed[row][col]:
        return
    revealed[row][col] = True
    if board[row][col] != ' ':
        return
    rows = len(board)
    cols = len(board[0])
    for i in range(max(0, row-1), min(rows, row+2)):
        for j in range(max(0, col-1), min(cols, col+2)):
            reveal_cell(board, revealed, i, j)

def print_board(board, revealed):
    rows = len(board)
    cols = len(board[0])
    print('    ' + '  '.join([str(i) for i in range(cols)]))
    print('  ' + '-' * (4*cols-1))
    for i in range(rows):
        print('{} | '.format(i), end='')
        for j in range(cols):
            if revealed[i][j]:
                print('{} | '.format(board[i][j]), end='')
            else:
                print('  | ', end='')
        print()
        print('  ' + '-' * (4*cols-1))

def play_game(rows, cols, num_mines):
    board = create_board(rows, cols)
    revealed = [[False]*cols for _ in range(rows)]
    place_mines(board, num_mines)

    game_over = False
    while not game_over:
        print_board(board, revealed)

        row = int(input("Enter row: "))
        col = int(input("Enter col: "))

        if board[row][col] == '*':
            revealed[row][col] = True
            print('Game Over! You hit a mine.')
            game_over = True
        else:
            reveal_cell(board, revealed, row, col)
            if all(board[i][j] != ' ' or revealed[i][j] for i in range(rows) for j in range(cols)):
                print_board(board, revealed)
                print('Congratulations! You won.')
                game_over = True

play_game(8, 8, 10)